README


This archive contains the distribution Installer,
version 0.903:

  What does it do? It installs stuff....

This software is copyright (c) 2013 by Torsten Raudssus.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.015.



Changes
==================================================
Changes from 1740-09-20 00:00:00 +0000 to present.
==================================================

------------------------------------------
version 0.903 at 2014-07-05 19:34:25 +0000
------------------------------------------

  Change: a1aa19b2bd19540b4cb768a2bf08254ee3ae1f79
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2014-07-05 21:34:00 +0000

    Adapting to new version of pluginbundle 

  Change: 6c0420dd4beb9b1ad78fd1ff787d7ced21bd8e06
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2014-05-20 21:58:33 +0000

    Removed Test::LoadAllModules 

  Change: f8485863e1df3a5ef35294bcc26dd8ca14baaf44
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2014-05-20 21:49:59 +0000

    Merge pull request #1 from crazedpsyc/master

    'Fix' the export.sh 

  Change: 89a87d6a81d865ad3f70aca13ae7e831193b6431
  Author: Michael Smith <crazedpsyc@duckduckgo.com>
  Date : 2014-04-11 17:17:28 +0000

    New .travis.yml 

  Change: 63128bae74bcdebce7e306068892e5530c396300
  Author: Michael Smith <crazedpsyc@duckduckgo.com>
  Date : 2014-04-11 17:16:04 +0000

    Avoid appending previously-empty variables with : 

  Change: fc65bf6007e47fa794975329746821fc66abcdd7
  Author: Michael Smith <crazedpsyc@duckduckgo.com>
  Date : 2014-04-07 22:32:19 +0000

    Nicer error message for an unknown Perl version 

  Change: bc167fb551ba5a1a8280b07a2f7098ae19719655
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-12-30 05:10:15 +0000

    Fixed 5.1 5.12 bug 

  Change: 3a7e64f62d496f9c5040a3ec8368e632efb01e73
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-12-30 04:21:15 +0000

    More documentation, some fixes, still some bugs... but it runs better
    for now 

  Change: 5b132b256ec398d47a1f662c4cfa927c6d281890
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-15 01:04:55 +0000

    Lost of changes.. still a bug 

  Change: c6cb0830e7b2a43345be8bab10087767d0485b12
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-11 05:06:57 +0000

    Added CURRENT_INSTALLER_ENV to export.sh 

  Change: 5aa4b60cf0cdf822123c05f28a09ba9332fdcc7b
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-11 04:37:07 +0000

    Another .travis.yml change... 

  Change: 5cedb318fa3fa623cafebdeab0c14083ef6e43a6
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-11 04:36:32 +0000

    Ok finally solved the base problem 

  Change: 23d5ba5ef7efaa46da509d2116af85ee56efebe9
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 06:43:07 +0000

    Renamed main binary to avoid collides with macosx users... <facepalm> 

  Change: d7eda661097ea966d637829fd1c3f8b22cf79e9f
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 05:12:36 +0000

    More stuff for fatpack.... I think ;) 

  Change: 5385efb2347b40459cf679f404c5ab443f0a6b90
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 04:10:00 +0000

    Added -e option 

  Change: 29948021573502452a19afac66ebbd8925cd5ac3
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 04:01:49 +0000

    Added reference to http://installer.pm on documentation 

  Change: 8ff22e8ea057c2e036201e6351fb3b911006f159
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 03:59:13 +0000

    Added fatpack for release process, removed old requirement 

  Change: 07094ff3aa499c65c18dfb1bf08c162254230a87
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 03:48:35 +0000

    Switched from Path::Tiny to Path::Class to be fatpackable 

  Change: ef5fae80f17a3e8b5f08bfad24e3c472f0222b7f
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 01:39:41 +0000

    Doc tuning 

  Change: 80c59120eb3814b32adf9ddf83b07d1ab2658d30
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 01:35:50 +0000

    Added URL support and a bit more documentation about OO usage 

  Change: 23a1aae1676d025aee32d37b8ab850cdb2fce0af
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 01:11:50 +0000

    More documentation and warnings and an OO example 

  Change: e0daed321dd2f1fb5b1a24b4f4e7e2ec71a1225f
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-09 00:15:41 +0000

    Python setup.py support and pip install option 

  Change: dc4e703a4fbe1f9267abd8b02f7d306cec187c21
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-08 23:21:02 +0000

    Small typo 

  Change: d4ecfeb6ec19f17041c299f7391a7d19f1ab20da
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-08 23:03:10 +0000

    More documentation 

  Change: 579be33ec206ac4816b95aebd17b6d4f192517ba
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-08 22:48:30 +0000

    Finished up file support 

  Change: 12ae0b9a6fe51c7d33f61d72ee0063f6bd4e895f
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-08 22:35:47 +0000

    Tuned requirements 

  Change: 2f58cb9a4c1b39209b55a2d9c204eea9d3beca5a
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-08 21:14:34 +0000

    Oh yeah... now with app 

  Change: 4ef23beebd0b76eaedaf2e5ba3e03ab0af859691
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-08 01:42:23 +0000

    Better output on failure 

  Change: d72d27adaeaaafdcce0de3e1987903a2681b8693
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-07 23:32:19 +0000

    Added .travis.yml 

  Change: 71106369905e65af72d15cf6e32e64f5105fc238
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-07 23:31:40 +0000

    More stuff... and making it now live 

  Change: 6b6d39460da02b82f93fd8182dd22a24bf736654
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-09-06 03:55:08 +0000

    initial commit 

================
End of releases.
================


lib/Installer/Target.pm
package Installer::Target;
BEGIN {
  $Installer::Target::AUTHORITY = 'cpan:GETTY';
}
# ABSTRACT: Currently running project
$Installer::Target::VERSION = '0.903';
use Moo;
use IO::All;
use IPC::Open3 ();
use Installer::Software;
use JSON_File;
use File::chdir;
use CPAN::Perl::Releases qw[perl_tarballs];
use CPAN;
use Path::Class;
use namespace::clean;

has output_code => (
  is => 'ro',
  lazy => 1,
  default => sub { sub {
    print @_, "\n";
  } },
);

has installer_code => (
  is => 'ro',
  required => 1,
);

has source_directory => (
  is => 'ro',
  predicate => 1,
);
has source => (
  is => 'ro',
  lazy => 1,
  default => sub { dir($_[0]->source_directory)->absolute },
);
sub source_path { dir(shift->source,@_) }
sub source_file { file(shift->source,@_) }

has target_directory => (
  is => 'ro',
  required => 1,
);
has target => (
  is => 'ro',
  lazy => 1,
  default => sub { dir($_[0]->target_directory)->absolute },
);
sub target_path { dir(shift->target,@_) }
sub target_file { file(shift->target,@_) }

has installer_dir => (
  is => 'ro',
  lazy => 1,
  default => sub { dir($_[0]->target,'installer') },
);

has software => (
  is => 'ro',
  lazy => 1,
  default => sub {{}},
);

has actions => (
  is => 'ro',
  lazy => 1,
  default => sub {[]},
);

has src_dir => (
  is => 'ro',
  lazy => 1,
  default => sub { dir($_[0]->target,'src') },
);

has log_filename => (
  is => 'ro',
  lazy => 1,
  default => sub { file($_[0]->installer_dir,'build.'.(time).'.log') },
);

has log_io => (
  is => 'ro',
  lazy => 1,
  default => sub { io($_[0]->log_filename) },
);

has meta => (
  is => 'ro',
  lazy => 1,
  default => sub {
    my ( $self ) = @_;
    tie(my %meta,'JSON_File',file($self->installer_dir,'meta.json')->absolute->stringify, pretty => 1);
    return \%meta;
  },
);

sub install_software {
  my ( $self, $software ) = @_;
  $self->software->{$software->alias} = $software;
  $software->installation;
  $self->meta->{software_packages_done} = [keys %{$self->software}];
  push @{$self->actions}, $software;
  $self->update_env;
  if (!defined $software->meta->{installed_export} && $software->has_export) {
    $self->install_export(ref $software->export eq 'ARRAY'
      ? @{$software->export}
      : $software->export
    );
    $software->meta->{installed_export} = 1;
  }
  if (!defined $software->meta->{installed_unset} && $software->has_unset) {
    $self->install_unset(ref $software->unset eq 'ARRAY'
      ? @{$software->unset}
      : $software->unset
    );
    $software->meta->{installed_unset} = 1;
  }
  $self->write_export;
  if (!defined $software->meta->{post_install} && $software->has_post_install) {
    $software->post_install->($software);
    $software->meta->{post_install} = 1;
  }
  $self->write_export;
}

sub install_url {
  my ( $self, $url, %args ) = @_;
  $self->install_software(Installer::Software->new(
    target => $self,
    archive_url => $url,
    %args,
  ));
}

sub install_file {
  my ( $self, $file, %args ) = @_;
  $self->install_software(Installer::Software->new(
    target => $self,
    archive => rel2abs(catfile($file)),
    %args,
  ));
}

sub install_perl {
  my ( $self, $perl_version, %args ) = @_;
  my $hashref = perl_tarballs($perl_version);
  die 'No such Perl version: '.$perl_version unless defined $hashref;
  my $src = 'http://www.cpan.org/authors/id/'.$hashref->{'tar.gz'};
  $self->install_software(Installer::Software->new(
    target => $self,
    archive_url => $src,
    testable => 1,
    custom_configure => sub {
      my ( $self ) = @_;
      $self->run($self->unpack_path,'./Configure','-des','-Dprefix='.$self->target_directory);
    },
    post_install => sub {
      my ( $self ) = @_;
      $self->log_print("Installing App::cpanminus ...");
      my $cpanm_filename = file($self->target->installer_dir,'cpanm');
      io($cpanm_filename)->print(io('http://cpanmin.us/')->get->content);
      chmod(0755,$cpanm_filename);
      $self->run(undef,$cpanm_filename,'-L',$self->target_path('perl5'),qw(
        App::cpanminus
        local::lib
        Module::CPANfile
      ));
    },
    export_sh => sub {
      my ( $self ) = @_;
      'eval $( perl -I'.$self->target_path('perl5','lib','perl5').' -Mlocal::lib=--deactivate-all )',
      'eval $( perl -I'.$self->target_path('perl5','lib','perl5').' -Mlocal::lib='.$self->target_path('perl5').' )'
    },
    %args,
  ));
}

#url "http://ftp.postgresql.org/pub/source/v9.3.0/postgresql-9.3.0.tar.bz2", with => {
#  pgport => 15700,
#};
sub install_postgres {
  my ( $self, $version, %args ) = @_;
  my $url = "http://ftp.postgresql.org/pub/source/v".$version."/postgresql-".$version.".tar.bz2";
  my %with = defined $args{with}
    ? %{delete $args{with}}
    : ();
  $with{pgport} = delete $args{port} if defined $args{port};
  my %users = defined $args{users}
    ? %{delete $args{users}}
    : ();
  my $pgdata = defined $args{data}
    ? dir(delete $args{data})->absolute->stringify
    : $self->target_path('pgdata')->absolute->stringify;
  my $logfile = defined $args{log}
    ? dir(delete $args{log})->absolute->stringify
    : $self->target_file('pgdata','postgresql.log');
  if (defined $args{superuser_with_db}) {
    my $superuser_with_db = delete $args{superuser_with_db};
    if (ref $superuser_with_db eq 'HASH') {
      for (keys %{$superuser_with_db}) {
        $users{$_} = {
          superuser => 1,
          dbs => [
            ref $superuser_with_db->{$_} eq 'ARRAY'
              ? @{$superuser_with_db->{$_}}
              : $superuser_with_db->{$_}
          ],
        };
      }
    } elsif (ref $superuser_with_db eq 'ARRAY') {
      for (@{$superuser_with_db}) {
        $users{$_} = {
          superuser => 1,
          dbs => [$_],
        };
      }
    } elsif (ref $superuser_with_db eq '') {
      $users{$superuser_with_db} = {
        superuser => 1,
        dbs => [$superuser_with_db],
      };
    } else {
      die "unknown how to handle ".(ref $superuser_with_db);
    }
  }
  my $post_install = delete $args{post_install};
  $self->install_software(Installer::Software->new(
    target => $self,
    archive_url => $url,
    %with ? ( with => \%with ) : (),
    export => [
      'PGDATA='.$pgdata,
      defined $with{pgport} ? ('PGPORT='.$with{pgport}, 'PGHOST=localhost') : (),
      defined $args{export} ? ( delete $args{export} ) : (),
    ],
    post_install => sub {
      my @post_install_args = @_;

      $_[0]->run(undef,'initdb');
      $_[0]->run(undef,'pg_ctl','-w','-l',$logfile,'start');

      for my $user (keys %users) {
        my @create_args = '-w';
        if ($users{$user}->{superuser}) {
          push @create_args, '-s';
        }
        $_[0]->run(undef,'createuser',@create_args,$user);
        if (defined $users{$user}->{dbs}) {
          for (@{$users{$user}->{dbs}}) {
            $_[0]->run(undef,'createdb','-O',$user,$_);
          }
        }
      }

      if (defined $post_install) {
        $post_install->(@post_install_args);
      }

      $_[0]->run(undef,'pg_ctl','stop');
    },
    %args,
  ));
}

sub install_cpanm {
  my ( $self, @modules ) = @_;
  $self->run(undef,'cpanm',@modules);
}

sub install_pip {
  my ( $self, @modules ) = @_;
  for (@modules) {
    $self->run(undef,'pip','install',$_);    
  }
}

sub install_run {
  my ( $self, @args ) = @_;
  $self->run($self->target,@args);
  push @{$self->actions}, {
    run => \@args,
  };
}

sub install_perldeps {
  my ( $self, $path, @args ) = @_;
  die "No source_directory or path given" unless defined $path || $self->has_source_directory;
  $self->run(undef,"cpanm","--installdeps",defined $path ? $path : $self->source_directory);
  $self->run(undef,"set");
}

sub install_dzildeps {
  my ( $self, $path, @args ) = @_;
  die "No source_directory or path given" unless defined $path || $self->has_source_directory;
  $self->run(undef,"cpanm","Dist::Zilla");
  my $dzil_dir = defined $path ? $path : $self->source_directory;
  $self->run($dzil_dir,qw( dzil authordeps | grep -v " " | cpanm ));
  $self->run($dzil_dir,qw( dzil listdeps | grep -v " " | cpanm ));
}

sub install_export {
  my ( $self, @args ) = @_;
  my @exports = defined $self->meta->{export}
    ? @{$self->meta->{export}}
    : ();
  for (@args) {
    my @new_exports;
    if (ref $_ eq 'CODE') {
      push @new_exports, $_->($self);
    } else {
      push @new_exports, $_;
    }
    for (@new_exports) {
      $self->log_print("Adding export ".$_);
      push @exports, $_;
    }
  }
  $self->meta->{export} = \@exports;
  $self->write_export;
}

sub install_unset {
  my ( $self, @args ) = @_;
  my @unsets = defined $self->meta->{unset}
    ? @{$self->meta->{unset}}
    : ();
  for (@args) {
    my @new_unsets;
    if (ref $_ eq 'CODE') {
      push @new_unsets, $_->($self);
    } else {
      push @new_unsets, $_;
    }
    for (@new_unsets) {
      $self->log_print("Adding export ".$_);
      push @unsets, $_;
    }
  }
  $self->meta->{unset} = \@unsets;
  $self->write_export;
}

sub patch_via_url {
  my ( $self, $path, $url, @args ) = @_;
  local $CWD = $path;
  $self->log_print("Fetching patch from $url into ".$path." ...");
  my $diff_name = $url;
  $diff_name =~ s/^https{0,1}//g;
  $diff_name =~ s/[^\w]+/_/g;
  $diff_name =~ s/^_+//g;
  $diff_name =~ s/_+$//g;
  $diff_name .= '.patch';
  io(file($path,$diff_name))->print(io($url)->get->content);
  $self->log_print("Applying patch as ".$diff_name." ...");
  $|=1;
  my $patch_log = "";
  my $pid = IPC::Open3::open3(my ( $in, $out ), undef, "patch",@args);
  print $in scalar io($diff_name)->slurp;
  close ($in);
  while(defined(my $line = <$out>)){
    $patch_log .= $line;
    chomp($line);
    $self->log($line);
  }
  waitpid($pid, 0);
  my $status = $? >> 8;
  if ($status) {
    print $patch_log;
    print "\n";
    print "     Command: patch ".join(" ",@args)."\n";
    print "in Directory: ".$path."\n\n";
    print "exited with status $status\n\n";
    die "Error on run ".$self->log_filename;
  }
}

sub run {
  my ( $self, $dir, @args ) = @_;
  $dir = $self->target_path unless $dir;
  local $CWD = "$dir";
  $self->log_print("Executing in $dir: ".join(" ",@args));
  $|=1;
  my $run_log = "";
  my $export_sh_filename = $self->target_file('export.sh')->absolute->stringify;
  my $prefix = "";
  if (-f $export_sh_filename) {
    my @export_sh_lines = io($export_sh_filename)->slurp;
    for my $line (@export_sh_lines) {
      $run_log .= $line;
      $prefix .= $line;
      chomp($line);
      $self->log($line);
    }
    $prefix .= "\n# Command ".("#" x 50)."\n";
  }
  my $shell_script = $prefix.join(" ",@args)."";
  my $pid = IPC::Open3::open3(my ( $in, $out ), undef, "/bin/sh -s");
  print $in $shell_script;
  close ($in);
  while(defined(my $line = <$out>)){
    $run_log .= $line;
    chomp($line);
    $self->log($line);
  }
  waitpid($pid, 0);
  my $status = $? >> 8;
  if ($status) {
    print $run_log;
    print "\n";
    print " Command:\n";
    print "\n".join(" ",@args)."\n\n";
    print " in Directory: ".$dir."\n";
    print "exited with status $status\n\n";
    print "\n";
    die "Error on run ".$self->log_filename;
  }
}

sub log {
  my ( $self, @line ) = @_;
  $self->log_io->append(join(" ",@line),"\n");
}

sub log_print {
  my ( $self, @line ) = @_;
  $self->log("#" x 80);
  $self->log("##");
  $self->log("## ",@line);
  my ($sec,$min,$hour,$mday,$mon,$year) = localtime(time);
  $self->log("## ",sprintf("%.2d.%.2d.%.4d %.2d:%.2d:%.2d",$mday,$mon,$year+1900,$hour,$min,$sec));
  $self->log("##");
  $self->log("#" x 80);
  $self->output_code->(@line);
}

our $current;

sub prepare_installation {
  my ( $self ) = @_;
  die "Target directory is a file" if -f $self->target;
  $current = $self;
  $self->target->mkpath unless -d $self->target;
  $self->installer_dir->mkpath unless -d $self->installer_dir;
  $self->src_dir->mkpath unless -d $self->src_dir;
  $self->log_io->print(("#" x 80)."\nStarting new log ".(time)."\n".("#" x 80)."\n\n");
  $self->meta->{last_run} = time;
  $self->meta->{preinstall_ENV} = \%ENV;
}

sub finish_installation {
  my ( $self ) = @_;
  $self->log_print("Done ".$self->log_filename);
  %ENV = %{$self->meta->{preinstall_ENV}};
  delete $self->meta->{preinstall_ENV};
  $current = undef;
}

sub installation {
  my ( $self ) = @_;
  $self->prepare_installation;
  $self->installer_code->($self);
  $self->finish_installation;
}

sub write_export {
  my ( $self ) = @_;
  my $export_filename = $self->target_file('export.sh');
  $self->log_print("Generating ".$export_filename." ...");
  my $export_sh = "#!/bin/sh\n#\n# Installer auto generated export.sh\n#\n".("#" x 60)."\n\n";
  $export_sh .= 'export CURRENT_INSTALLER_ENV='.$self->target_path->stringify."\n";
  if (defined $self->meta->{unset} && @{$self->meta->{unset}}) {
    $export_sh .= '# custom unsets'."\n";
    for (@{$self->meta->{unset}}) {
      $export_sh .= 'unset '.$_."\n";
    }
  }
  if (defined $self->meta->{PATH} && @{$self->meta->{PATH}}) {
    $export_sh .= 'export PATH="'.join(':',@{$self->meta->{PATH}}).'${PATH+:}$PATH"'."\n";
  }
  if (defined $self->meta->{LD_LIBRARY_PATH} && @{$self->meta->{LD_LIBRARY_PATH}}) {
    $export_sh .= 'export LD_LIBRARY_PATH="'.join(':',@{$self->meta->{LD_LIBRARY_PATH}}).'${LD_LIBRARY_PATH+:}$LD_LIBRARY_PATH"'."\n";
  }
  if (defined $self->meta->{C_INCLUDE_PATH} && @{$self->meta->{C_INCLUDE_PATH}}) {
    $export_sh .= 'export C_INCLUDE_PATH="'.join(':',@{$self->meta->{C_INCLUDE_PATH}}).'${C_INCLUDE_PATH+:}$C_INCLUDE_PATH"'."\n";
  }
  if (defined $self->meta->{MANPATH} && @{$self->meta->{MANPATH}}) {
    $export_sh .= 'export MANPATH="'.join(':',@{$self->meta->{MANPATH}}).'${MANPATH+:}$MANPATH"'."\n";
  }
  if (defined $self->meta->{export} && @{$self->meta->{export}}) {
    $export_sh .= '# custom exports'."\n";
    for (@{$self->meta->{export}}) {
      $export_sh .= 'export '.$_."\n";
    }
  }
  $export_sh .= "\n";
  for (@{$self->meta->{software_packages_done}}) {
    my $software = $self->software->{$_};
    if ($software->has_export_sh) {
      my @lines = $software->export_sh->($software);
      $export_sh .= "# export.sh addition by ".$software->alias."\n";
      $export_sh .= join("\n",@lines)."\n\n";
    }
  }
  $export_sh .= ("#" x 60)."\n";
  io($export_filename)->print($export_sh);
  chmod(0755,$export_filename);
}

sub update_env {
  my ( $self ) = @_;
  my %seen = defined $self->meta->{seen_dirs}
    ? %{$self->meta->{seen_dirs}}
    : ();
  if (!$seen{'bin'} and -e $self->target_path('bin')) {
    my @bindirs = defined $self->meta->{PATH}
      ? @{$self->meta->{PATH}}
      : ();
    my $bindir = $self->target_path('bin')->absolute->stringify;
    push @bindirs, $bindir;
    $self->meta->{PATH} = \@bindirs;
    $seen{'bin'} = 1;
  }
  if (!$seen{'man'} and -e $self->target_path('man')) {
    my @mandirs = defined $self->meta->{MANPATH}
      ? @{$self->meta->{MANPATH}}
      : ();
    my $mandir = $self->target_path('man')->absolute->stringify;
    push @mandirs, $mandir;
    $self->meta->{MANPATH} = \@mandirs;
    $seen{'man'} = 1;
  }
  if (!$seen{'lib'} and -e $self->target_path('lib')) {
    my @libdirs = defined $self->meta->{LD_LIBRARY_PATH}
      ? @{$self->meta->{LD_LIBRARY_PATH}}
      : ();
    my $libdir = $self->target_path('lib')->absolute->stringify;
    push @libdirs, $libdir;
    $self->meta->{LD_LIBRARY_PATH} = \@libdirs;
    $seen{'lib'} = 1;
  }
  if (!$seen{'include'} and -e $self->target_path('include')) {
    my @libdirs = defined $self->meta->{C_INCLUDE_PATH}
      ? @{$self->meta->{C_INCLUDE_PATH}}
      : ();
    my $libdir = $self->target_path('include')->absolute->stringify;
    push @libdirs, $libdir;
    $self->meta->{C_INCLUDE_PATH} = \@libdirs;
    $seen{'include'} = 1;
  }
  $self->meta->{seen_dirs} = \%seen;
}

1;

__END__

=pod

=head1 NAME

Installer::Target - Currently running project

=head1 VERSION

version 0.903

=head1 DESCRIPTION

You should use this through the command L<installto>.

B<TOTALLY BETA, PLEASE TEST :D>

=head1 SUPPORT

IRC

  Join #cindustries on irc.quakenet.org. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-installer
  Pull request and additional contributors are welcome

Issue Tracker

  http://github.com/Getty/p5-installer/issues

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Torsten Raudssus.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
